device/core/src/sak_authentication_provider.ts

```diff
@@ -54,46 +62,50 @@ export class SharedAccessKeyAuthenticationProvider extends EventEmitter implemen
     if (this._shouldRenewToken()) {
-      this._renewToken();
+      this._renewToken(callback);
+    } else {
+      callback(null, this._credentials);
     }
-
-    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_003: [The `getDeviceCredentials` should call its callback with a `null` first parameter and a `TransportConfig` object as a second parameter, containing the latest valid token it generated.]*/
-    callback(null, this._credentials);
   }
 
   private _shouldRenewToken(): boolean {
     const currentTimeInSeconds = Math.floor(Date.now() / 1000);
     return (this._currentTokenExpiryTimeInSeconds - currentTimeInSeconds) < this._tokenRenewalMarginInSeconds;
   }
 
-  private _renewToken(): void {
+  private _renewToken(callback?: (err: Error, credentials?: TransportConfig) => void): void {
     if (this._renewalTimeout) {
       clearTimeout(this._renewalTimeout);
     }
 
-    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_009: [Every token shall be created with a validity period of `tokenValidTimeInSeconds` if specified when the constructor was called, or 1 hour by default.]*/
-    const newExpiry =  Math.floor(Date.now() / 1000) + this._tokenValidTimeInSeconds;
-
-    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_010: [Every token shall be created using the `azure-iot-common.SharedAccessSignature.create` method and then serialized as a string, with the arguments to the create methods being:
-    ```
-    resourceUri: <IoT hub host>/devices/<deviceId>
-    keyName: the `SharedAccessKeyName` parameter of the connection string or `null`
-    key: the `SharedAccessKey` parameter of the connection string
-    expiry: the expiration time of the token, which is now + the token validity time, formatted as the number of seconds since Epoch (Jan 1st, 1970, 00:00 UTC).
-    ```]*/
     let resourceString = this._credentials.host + '/devices/' + this._credentials.deviceId;
     if (this._credentials.moduleId) {
       resourceString += '/modules/' + this._credentials.moduleId;
     }
     const resourceUri = encodeUriComponentStrict(resourceString);
-    const sas = SharedAccessSignature.create(resourceUri, this._credentials.sharedAccessKeyName, this._credentials.sharedAccessKey, newExpiry);
-    this._credentials.sharedAccessSignature = sas.toString();
 
-    const nextRenewalTimeout = (this._tokenValidTimeInSeconds - this._tokenRenewalMarginInSeconds) * 1000;
-    this._renewalTimeout = setTimeout(() => this._renewToken(), nextRenewalTimeout);
-    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_005: [Every time a new token is created, the `newTokenAvailable` event shall be fired with the updated credentials.]*/
-    this.emit('newTokenAvailable', this._credentials);
+    this._signatureProvider.sign(this._credentials.sharedAccessKeyName, resourceUri, (err, sas) => {
+      if (err) {
+        this.emit('error', err);
```

Comment:
>this.emit('error', err); [](start = 8, length = 24)

i'd probably say to emit the error only if there is no callback?

Reply:
Sure.